src/server/raftkv.rs

```diff
@@ -633,3 +665,65 @@ impl<S: Snapshot> EngineIterator for RegionIterator<S> {
         RegionIterator::value(self)
     }
 }
+
+#[derive(Clone)]
+pub struct ReplicaReadLockChecker {
+    concurrency_manager: ConcurrencyManager,
+}
+
+impl ReplicaReadLockChecker {
+    pub fn new(concurrency_manager: ConcurrencyManager) -> Self {
+        ReplicaReadLockChecker {
+            concurrency_manager,
+        }
+    }
+
+    pub fn register<E: KvEngine + 'static>(self, host: &mut CoprocessorHost<E>) {
+        host.registry
+            .register_read_index_observer(1, BoxReadIndexObserver::new(self));
+    }
+}
+
+impl Coprocessor for ReplicaReadLockChecker {}
+
+impl ReadIndexObserver for ReplicaReadLockChecker {
+    fn on_step(&self, msg: &mut eraftpb::Message) -> bool {
+        if msg.get_msg_type() != MessageType::MsgReadIndex {
+            return false;
+        }
+        assert_eq!(msg.get_entries().len(), 1);
+        let mut rctx = ReadIndexContext::parse(msg.get_entries()[0].get_data()).unwrap();
```

Comment:
Is it possible it's from an old version TiKV?

Reply:
Old version TiKV uses a UUID as the context. It can handle the old format.

Reply:
I see.